### PR TITLE
Ginkgo wrapper for deal.ii

### PR DESCRIPTION
**Pull request by pratikvn**
Hello,

This pull request is concerned with the issue #7183 and aims to add a wrapper for the [ginkgo](https://github.com/ginkgo-project/ginkgo) library in deal.ii. 

You can also see `/docs/readme.html` for instructions on how to install ginkgo and the necessary flags or even  [ginkgo](https://github.com/ginkgo-project/ginkgo)'s repository. 

As discussed, I copy over the matrix and rhs, convert them to the required format , solve and then pass the solution back in deal.ii's format. For simplicity and as a first pass, I only implement the CG solver with no preconditioners (interfaces to these will be added in future pull requests). 

The tests in `/tests/ginkgo/` demonstrate how to use ginkgo as a solver. You can see that you currently have three choices to choose where you want to solve your system. 

1. `omp` to solve on the CPU within the OpenMP paradigm (Ginkgo needs to be built with `-DBUILD_OMP=on`)

2. `reference` to solve on the CPU with a serial reference implementation (Ginkgo needs to be built with `-DBUILD_REFERENCE=on`)

3. `cuda` to solve on the GPU. (Ginkgo needs to be built with `-DBUILD_CUDA=on`) and an appropriate CUDA version needs to be installed. Additionally, a cuda device needs to be available to solve on.

These choices are called `executors` and you can choose only one of them. Additionally ,the `cuda` executor needs to have a cpu executor (either the `omp`/`reference` for now ) as a `master`. 

I am not yet clear about your CI system, so please advise on that.
 
If you have any suggestions / improvements , I would be happy to add them.  

Thank you,
Pratik.

### CODE REVIEW

File: doc/external-libs/ginkgo.html [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+        Installing Ginkgo is quite simple. Ginkgo currently uses Cmake and hence one can use
+        the following commands to easily install Ginkgo. For different dependencies and compatible
+        libraries, please refer to <a href="https://github.com/ginkgo-project/ginkgo"
+                                      target="_top">Ginkgo</a>.
+        The different flags that can be added are:
+        <code> -DBUILD_REFERENCE={ON,OFF}:</code> Builds the reference single thread implementation
+        used to check the more sophisticated implementations of the GPU/CPU.
+        <code> -DBUILD_CUDA={ON,OFF}:</code> Builds the GPU implementation specifically the NVIDIA
+        cuda implementations. This needs CUDA to be installed on the machine.
+        <code> -DBUILD_OMP={ON,OFF}:</code> Builds the multi-threaded OpenMP implementations.
```
Comment (davydden) [inline doc/external-libs/ginkgo.html]: this is relevant to https://github.com/dealii/dealii/issues/7300

I am not sure one should recommend disabling `openmp`. I had bad experience when `openmp` was enabled in `openblas` together with `tbb` active in `deal.II` (and I was NOT calling `blas` functions from TBB-parallel part of `dealii`/user code). But this could be specific to `openblas` implementation of multithreading...

Could you maybe tests that you don't see any slow-downs when `omp` is enabled in `gingo`?

Comment (pratikvn) [reply doc/external-libs/ginkgo.html]: Thank you for your comments. Hm, this is interesting. Do you suggest any specific timed examples or tests that you have which I can use ?

Comment (pratikvn) [reply doc/external-libs/ginkgo.html]: Thanks. I have tested this. I do not observe any slowdown and I see an appropriate scaling for large enough problem sizes. I do not expect this to be a problem in the future (as long as the interface to ginkgo remains the same) as well because during the solution phase when deal.ii gives the control over to ginkgo, there shouldn't be any threads "working" within deal.ii which might interfere with the Omp threads spawned by Ginkgo.

Comment (davydden) [reply doc/external-libs/ginkgo.html]: Thanks for checking this

**Comment by masterleinad** [comment]
The `gingko/solver.debug` test case fails for me with
```
  what():  ../core/device_hooks/common_kernels.inc.cpp:189: feature initialize is part of the GKO_HOOK_MODULE module, which is not compiled on this system
```

**Comment by pratikvn** [comment]
Can you please confirm the following things:
1. That you have compiled ginkgo with `-DBUILD_OMP=on` and `-DBUILD_REFERENCE=on` ? In the test `dealii/tests/ginkgo/solver.*/` the default executor is `omp`. If you have not compiled with  `-DBUILD_OMP=on` you will not have the appropriate functions. You can also try commenting out the `omp` executor and uncommenting the next line to use the `reference` executor instead.

2. If you have checked out the latest develop branch of ginkgo from github ? 

3. If you do `ldd` on your library object (`libdeal_II.*so*`), does it show that it is linked to all four ginkgo libraries, `libginkgo.so`, `libginkgo_reference.so` , `libginkgo_cuda.so` and `libginkgo_omp.so` ?

**Comment by masterleinad** [comment]
I only tried what a typical user would do, i.e. building `ginkgo` with a default configuration.
Then, unsurprisingly, only the first one is found.
```
	libginkgo.so => /mnt/data/darndt/ginkgo/lib/libginkgo.so (0x00007f7d53048000)
	libginkgo_omp.so => not found
	libginkgo_cuda.so => not found
	libginkgo_reference.so => not found
```

If I enable `REFERENCE` and `OMP` in `ginkgo` the test works.

I just wanted to point out that you might want to specify more requirements when configuring `ginkgo` in `deal.II`.

**Comment by pratikvn** [comment]
By default, `-DBUILD_REFERENCE` should be `on`. So, without any specific flags, it should build with reference kernels. Additionally, even if you do not build with `BUILD_CUDA` or `BUILD_OMP`, it should build with hooks for all the `libginkgo_*.so` should be in `/mnt/data/darndt/ginkgo/lib/` for you. I am surprised that only the first one is found. 

But thanks for pointing this out. Maybe I can add a note saying that you should configure with `-BUILD_OMP` ? 

Also maybe for the test, I should make the default executor `reference` for now ? I plan to add tests for all the executors which the library has been compiled for in the future.

**Comment by pratikvn** [comment]
Can someone please have a look at this ? As requested, only new files are added and no modifications to any of the existing deal.ii files has been done.

Currently, only the CG solver with no preconditioner interface and only with the CSR matrix has been added. This was done to prevent the pull request from being too large and to make the review process easier. Would you prefer that I also add the other solvers, support for preconditioners and the matrix formats within this pull request ? Or would you prefer a new pull request for those ?

**Comment by jppelteret** [comment]
Hi @pratikvn! Thanks for your patience with this PR. I'll do my best to review it this coming weekend, but in the mean time I have a couple of cosmetic suggestions and one question after a very quick glance through it:

1. Please correct the copyright dates in all of the new files.
2. There are a couple of places in the documentation when you write `deal.II` as `deal.ii`. May you please correct this.
3. Is it possible to create a constructor to the concrete solver classes that looks like this
```
    SolverCG(SolverControl &                solver_control,
             const AdditionalData &         data = AdditionalData());
```
i.e. without explicitly passing a Ginko executor? My thought is that it might be possible to choose the best performing executor for the user based a the detected feature set. I recognise that this would probably exclude the opportunity to use the GPU because this likely requires a more intricate construction. But based on the commented out lines in your test
```
    // std::shared_ptr<gko::Executor> exec = gko::CudaExecutor::create(0,
    //                                                               gko::OmpExecutor::create());
    // std::shared_ptr<gko::Executor> exec = gko::OmpExecutor::create();
    std::shared_ptr<gko::Executor> exec = gko::ReferenceExecutor::create();
```
it appears to me that it might be possible to automatically choose the most performant option for the user should they wish to use this option.

**Comment by pratikvn** [comment]
Thank you for your comments. I have fixed the copyright and the doc issues. 

Regarding the automatic selection of the executor, that is a good idea. I will create an issue about this in deal.II repo, because I think it needs a bit of discussion as to how you would choose the executor. It is not yet completely clear to me.

File: cmake/configure/configure_2_ginkgo.cmake [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+##
+## ---------------------------------------------------------------------
+
+#
+# Configuration for the ginkgo library:
+#
+
+MACRO(FEATURE_GINKGO_ERROR_MESSAGE)
+  MESSAGE(FATAL_ERROR "\n"
+    "Could not find ginkgo and supporting libraries!\n"
```
Comment (jppelteret) [inline cmake/configure/configure_2_ginkgo.cmake]: `ginkgo` -> `Ginkgo`

File: cmake/configure/configure_2_ginkgo.cmake [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+    "Please ensure that the libraries are installed on your computer.\n"
+    "If the libraries are not at a default location, either provide some hints\n"
+    "for the autodetection:\n"
+    "    $ GINKGO_DIR=\"...\" cmake <...>\n"
+    "    $ cmake -DGINKGO_DIR=\"...\" <...>\n"
+    "or set the relevant variables by hand in ccmake.\n"
+    "Relevant hints for GINKGO are GINKGO_DIR\n"
```
Comment (jppelteret) [inline cmake/configure/configure_2_ginkgo.cmake]: "The relevant hints for GINKGO are GINKGO_DIR.\n"

File: cmake/modules/FindGINKGO.cmake [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+#
+# This module exports
+#
+#   GINKGO_INCLUDE_DIRS
+#
+
+SET(GINKGO_DIR "" CACHE PATH "An optional hint to a GINKGO installation")
+SET_IF_EMPTY(GINKGO_DIR "$ENV{GINKGO_DIR}")
+
+DEAL_II_FIND_LIBRARY(GINKGO_LIBRARY
```
Comment (jppelteret) [inline cmake/modules/FindGINKGO.cmake]: We typically put these details inside a module, which in this case should be under `cmake/modules/FindGINKO.cmake`. Please see this example ([module file](https://github.com/dealii/dealii/blob/master/cmake/modules/FindMETIS.cmake); [configure file](https://github.com/dealii/dealii/blob/master/cmake/configure/configure_2_metis.cmake)) to see how its normally structured.

File: doc/external-libs/ginkgo.html [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+  </head>
+
+  <body>
+
+    <h1>Interfacing <acronym>deal.II</acronym> to Ginkgo</h1>
+
+    <p>
+        <a href="https://github.com/ginkgo-project/ginkgo"
+           target="_top">Ginkgo</a> is a
+        Numerical linear algebra software package that provides its users with
```
Comment (jppelteret) [inline doc/external-libs/ginkgo.html]: A numerical...

File: doc/external-libs/ginkgo.html [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+        highly optimized fine grid level linear algebra operations. It currently has
+        the capability to solve on the CPU with the support of OpenMP and on the GPU
+        with NVIDIA's cuda libraries. For some operations it uses NVIDIA's own libraries, CuSparse
```
Comment (jppelteret) [inline doc/external-libs/ginkgo.html]: cuda -> CUDA

Both here and further below.

File: doc/external-libs/ginkgo.html [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+        and CuBLAS but some routines also have self implemented cuda functions. See the documentation
```
Comment (jppelteret) [inline doc/external-libs/ginkgo.html]: and CuBLAS, but some

(comma is missing)

Comment (jppelteret) [inline doc/external-libs/ginkgo.html]: It would be good to clarify if the phrase "the documentation" refers to Ginko's official documentation, or that of the wrappers within the deal.II library.

Also, can you hyperlink the phrase "the documentation" to relevant documentation?

File: doc/external-libs/ginkgo.html [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+    <p>
+      <acronym>deal.II</acronym> has wrapper classes to the linear algebra
+      parts of Ginkgo that provide almost the
+      same interfaces as the built-in <acronym>deal.II</acronym> linear algebra operations.
+    </p>
+
+    <h2>Installing <acronym>deal.II</acronym> with Ginkgo</h2>
+
+    <p>
+        During the cmake configuration, the following flags should be specified
```
Comment (jppelteret) [inline doc/external-libs/ginkgo.html]: cmake -> CMake

Please correct the spelling below as well.

File: doc/external-libs/ginkgo.html [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+        <code>-DDEAL_II_WITH_GINKGO=ON</code>, to enable Ginkgo and the <code>-DGINKGO_DIR=DIR</code>
+        can be set with the path where Ginkgo has been installed.
```
Comment (jppelteret) [inline doc/external-libs/ginkgo.html]: I would suggest that this whole paragraph needs to be restructured slightly to be clear. May I suggest the following:

```
During the CMake configuration, the following flags should be specified:
<ul>
  <li> <code>-DDEAL_II_WITH_GINKGO=ON</code>, to enable the Ginkgo library
  <li> <code>-DGINKGO_DIR=<path_to_ginko></code>, to specify the path where Ginkgo has been installed.
</ul>
```

File: doc/external-libs/ginkgo.html (hunk 1 above)
Comment (jppelteret) [inline doc/external-libs/ginkgo.html]: multi-threaded -> multithreaded

Comment (jppelteret) [inline doc/external-libs/ginkgo.html]: Here you can also use the `<ul>` `</ul>` tags to create a list, as was suggested my the previous comment.

File: doc/external-libs/ginkgo.html [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+        libraries, please refer to <a href="https://github.com/ginkgo-project/ginkgo"
+                                      target="_top">Ginkgo</a>.
+        The different flags that can be added are:
+        <code> -DBUILD_REFERENCE={ON,OFF}:</code> Builds the reference single thread implementation
+        used to check the more sophisticated implementations of the GPU/CPU.
+        <code> -DBUILD_CUDA={ON,OFF}:</code> Builds the GPU implementation specifically the NVIDIA
+        cuda implementations. This needs CUDA to be installed on the machine.
+        <code> -DBUILD_OMP={ON,OFF}:</code> Builds the multi-threaded OpenMP implementations.
+
+      <pre>
```
Comment (jppelteret) [inline doc/external-libs/ginkgo.html]: Please add some sort of a description as to what the next set of commands do (i.e. add a descriptive line just above them).

File: doc/external-libs/ginkgo.html [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+
+	git clone https://github.com/ginkgo-project/ginkgo.git
+  mkdir build; cd build 
+  cmake -DBUILD_REFERENCE=on/off -DBUILD_CUDA=on/off -DBUILD_OMP=on/off -DCMAKE_INSTALL_PREFIX=/path/to/install/ -DCMAKE_BUILD_TYPE=Debug
+	make install
+      </pre>
+    </p>
+
+    <h4>Running Ginkgo tests</h4>
```
Comment (jppelteret) [inline doc/external-libs/ginkgo.html]: the Ginko

File: doc/external-libs/ginkgo.html [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+
+    <p>
+        Ginkgo uses <code>gtests</code> to run unit tests. The reference implementations are
+        non-optimized versions to which the OpenMP and cuda versions are compared with for
+        correctness.
+      <pre>
```
Comment (jppelteret) [inline doc/external-libs/ginkgo.html]: Please add a introductory line above this command. e.g. "The following command, executed in Ginko's build directory, launches its testsuite."

File: doc/external-libs/ginkgo.html [hunk 13]
```diff
@@ -0,0 +1,98 @@
...
+      To run only a specific test for more details, from the build directory, run
+      <pre>
+	./path/to/test
+      </pre>
+    </p>
+
+    <h5>NOTES </h5>
+    <p>
+        Currently, Ginkgo needs to be compiled in "debug mode", due to some problems with
+        a dependent <code>gflags</code> dependency.
```
Comment (jppelteret) [inline doc/external-libs/ginkgo.html]: > dependent... dependency

This is quite awkward. Could you try to rephrase this to make more clear what it means?

File: doc/readme.html [hunk 14]
```diff
@@ -423,6 +423,22 @@ <h4>Optional interfaces to other software packages</h4>
                 about compatibility and configuration can be found <a href=external-libs/cuda.html target="body">here</a>.
             </p>
 
+            <dt><a name="Ginkgo"/>
+                <a href="https://github.com/ginkgo-project/ginkgo" target="_top">Ginkgo</a></dt>
+            <dd>
+                <p>
+                    <a href="https://github.com/ginkgo-project/ginkgo" target="_top">Ginkgo</a>
+                    is a numerical linear algebra library with highly optimized kernels
+                    for many core architectures with focus on fine level parallelism.
```
Comment (jppelteret) [inline doc/readme.html]: architectures, with a focus

(missing comma and "a")

File: doc/readme.html [hunk 15]
```diff
@@ -423,6 +423,22 @@ <h4>Optional interfaces to other software packages</h4>
...
+                    It allows for easy switching of the executing paradigm (cuda, omp, .etc)
```
Comment (jppelteret) [inline doc/readme.html]: (CUDA, OpenMP, etc.)

File: doc/readme.html [hunk 16]
```diff
@@ -423,6 +423,22 @@ <h4>Optional interfaces to other software packages</h4>
...
+                    while providing a multitude of linear solvers and preconditioners
+                    and extension to other linear operators with no changes required in the code,
```
Comment (jppelteret) [inline doc/readme.html]: > with no changes required in the code

This, as far as I understand, is not quite true. You've got the different `gko::Executor` that need to be configured. I think that the word "no" should be substituted by "minimal".

File: doc/readme.html [hunk 17]
```diff
@@ -423,6 +423,22 @@ <h4>Optional interfaces to other software packages</h4>
...
+                    pass <code>-DGINKGO_DIR=/path/to/ginkgo</code> to the
+                    deal.II CMake call. For more detailed instructions, one can refer to
```
Comment (jppelteret) [inline doc/readme.html]: > changes required in the code, pass <code>-DGINKGO_DIR=/path/to/ginkgo</code> to the deal.II CMake call.

Two distinct concepts have been merged here. I think that you mean to say something like

```
changes required in the code. To enable Ginko, pass <code>-DGINKGO_DIR=/path/to/ginkgo</code> to CMake when configuring deal.II.
```

File: cmake/modules/FindGINKGO.cmake (hunk 4 above)
Comment (jppelteret) [reply cmake/modules/FindGINKGO.cmake]: Whoops sorry, I made a mistake here. I didn't notice that I was reviewing `FindGINKO.cmake`. Apologies :-/

**Comment by pratikvn** [comment]
Thank you for the comments @jppelteret.

File: include/deal.II/lac/ginkgo_solver.h [hunk 18]
```diff
@@ -0,0 +1,188 @@
...
+   */
+  class SolverBase
+  {
+  public:
+    /**
+     * Constructor. Takes the solver control object and the executor. The
+     * executor defines the paradigm where the solution is computed. You can
+     * choose between an `omp` executor, which would solve the linear system
+     * using OpenMP threading on a single machine or a `gpu` executor, which
+     * would solve the system on a GPU, only if it is available. You also have a
```
[the remark this answers is not in the document]
Comment (pratikvn) [reply include/deal.II/lac/ginkgo_solver.h]: A run-time error would be thrown stating that the function in question has not been implemented.

File: include/deal.II/lac/ginkgo_solver.h [hunk 19]
```diff
@@ -0,0 +1,188 @@
...
+    std::shared_ptr<gko::log::Convergence<>> convergence_logger;
+
+    /**
+     * The ginkgo combined factory object is used to create a combined stopping
+     * criterion to be passed to the solver.
+     */
+    std::shared_ptr<gko::stop::Combined::Factory> combined_factory;
+
+    /**
+     * The execution paradigm in Ginkgo. The choices lie between `omp`, `gpu`
```
[the remark this answers is not in the document]
Comment (pratikvn) [reply include/deal.II/lac/ginkgo_solver.h]: I am not quite sure I understand what you mean here.

File: include/deal.II/lac/ginkgo_solver.h [hunk 20]
```diff
@@ -0,0 +1,188 @@
...
+     * and `ref` and more details can be found in Ginkgo's documentation.
+     */
+    std::shared_ptr<gko::Executor> executor;
+
+  private:
+    /**
+     * Initialize the ginkgo logger object with event masks.
```
[the remark this answers is not in the document]
Comment (pratikvn) [reply include/deal.II/lac/ginkgo_solver.h]: I have added a link and [here](https://github.com/ginkgo-project/ginkgo/blob/develop/core/log/logger.cpp) and [here](https://github.com/ginkgo-project/ginkgo/blob/develop/include/ginkgo/core/log/logger.hpp) you can find more details. Essentially, the masks define which you events you want to log and a list of masks are available at the link above.

File: source/lac/ginkgo_solver.cc [hunk 21]
```diff
@@ -0,0 +1,273 @@
...
+    : solver_control(solver_control)
+    , executor(executor)
+  {}
+
+  void
+  SolverBase::ginkgo_log()
+  {
+    // Add the logger object. See the different masks available in Ginkgo's
+    // documentation
+    convergence_logger = gko::log::Convergence<>::create(
```
[the remark this answers is not in the document]
Comment (pratikvn) [reply source/lac/ginkgo_solver.cc]: I could but I wanted to emphasize on the logging capability and if anyone wanted to add more information for certain solvers, I can see it also called from the derived classes.

File: include/deal.II/lac/ginkgo_solver.h (hunk 19 above)
[the remark this answers is not in the document]
Comment (masterleinad) [reply include/deal.II/lac/ginkgo_solver.h]: Is the execution paradigm's name really `omp`, `gpu` or is it `gko::omp`,`gko::gnu` (or similar)?

File: include/deal.II/lac/ginkgo_solver.h [hunk 22]
```diff
@@ -0,0 +1,188 @@
...
+     */
+    void
+    ginkgo_log();
```
[the remark this answers is not in the document]
Comment (bangerth) [reply include/deal.II/lac/ginkgo_solver.h]: As is the tree, by the way :-)

File: include/deal.II/lac/ginkgo_solver.h (hunk 19 above)
[the remark this answers is not in the document]
Comment (pratikvn) [reply include/deal.II/lac/ginkgo_solver.h]: You create an executor only in the user code. See a simple solver example [here](https://github.com/ginkgo-project/ginkgo/blob/develop/examples/simple_solver/simple_solver.cpp). 

Ah, maybe you meant add it as `gko::ReferenceExecutor`,  `gko::OmpExecutor`... in the documentation for clarity ? I will do that.

File: include/deal.II/lac/ginkgo_solver.h (hunk 22 above)
[the remark this answers is not in the document]
Comment (pratikvn) [reply include/deal.II/lac/ginkgo_solver.h]: And logging the tree was not an intended pun  :-)

[the remark this answers is not in the document]
Comment (bangerth) [reply include/deal.II/lac/ginkgo_solver.h]: Oh, what a fun pun!

**Comment by pratikvn** [comment]
@jppelteret ,Thank you very much for the detailed review. Hopefully, everything except the feature set detection should be fixed. We will soon add those variables to our configure file so that deal.II can use it. If we can merge this, then for the next wave of additions, that will be added as well.

**Comment by jppelteret** [comment]
@pratikvn You're very welcome. I just need a day or two to be able to look through the changes you've made. In the mean time, would you mind please rebasing the branch such that it has a few logical commits (28 is a bit too many :-). Maybe having one for the CMake stuff, one for the implementation and one for the test might be the appropriate division. After that, we can set the tester to run on your PR!

I've also just noticed that you've not added a changelog entry. Without question your efforts need to be rewarded with one! Please add one to [`doc/news/major`](https://github.com/dealii/dealii/tree/master/doc/news/changes/major). You can use an existing entry as a template (but please try to stick to about 80 characters on a line).

Thanks!

File: include/deal.II/lac/ginkgo_solver.h (hunk 22 above)
[the remark this answers is not in the document]
Comment (jppelteret) [reply include/deal.II/lac/ginkgo_solver.h]: Sigh... My eyes and brain had a serious disagreement here! Sorry :-D

File: cmake/configure/configure_2_ginkgo.cmake [hunk 23]
```diff
@@ -0,0 +1,33 @@
...
+    )
+ENDMACRO()
+
```
[the remark this answers is not in the document]
Comment (pratikvn) [reply cmake/configure/configure_2_ginkgo.cmake]: We have added some CMake variables which should be exported ginkgo is installed. I have defined some macros in `configure_2_ginkgo.cmake` similar to the one in adolc. I guess we could filter out some tests, for actual usage we throw exceptions which should display the error properly if some module is not compiled and some function for that module is called.

[the remark this answers is not in the document]
Comment (jppelteret) [reply cmake/configure/configure_2_ginkgo.cmake]: Great, I think that this might help with #7605.

File: doc/readme.html [hunk 24]
```diff
@@ -423,6 +423,22 @@ <h4>Optional interfaces to other software packages</h4>
...
             </p>
 
+            <dt><a name="Ginkgo"/>
+                <a href="https://ginkgo-project.github.io/" target="_top">Ginkgo</a></dt>
+            <dd>
+                <p>
+                    <a href="https://ginkgo-project.github.io/" target="_top">Ginkgo</a>
+                    is a numerical linear algebra library with highly optimized kernels
+                    for many core architectures with a focus on fine level parallelism.
+                    It allows for easy switching of the executing paradigm (CUDA, OpenMP, .etc)
```
[the remark this answers is not in the document]
Comment (pratikvn) [reply doc/readme.html]: Done. Thanks!

**Comment by jppelteret** [comment]
I think that this PR is in good form now, and its author(s) have been patient in waiting for its review. Since there have been no other comments thus far I feel that its ready to be added to the library.

Thanks @pratikvn and the other Ginkgo developers for your contribution to the deal.II library! We look forward to future PR', and seeing how the wrappers to the other solvers and the other functionality of the Ginkgo library shape up.
